### ISSUE TITLE
Compilation error when using Sundials and MPICH

### ISSUE THREAD

**Issue by jppelteret** [dealii/dealii#12914]
From https://github.com/spack/spack/issues/27183. It looks like there are some troubles when the Sundials wrappers are compiled using MPICH.

Salient lines from build output:
```
     1685    In file included from /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-src/source/sundials/n_vector.cc:20:
     1686    /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-src/include/deal.II/sundials/n_vector.templates.h: In instantiation o
             f 'void* dealii::SUNDIALS::internal::NVectorOperations::get_communicator_as_void_ptr(N_Vector) [with VectorType = dealii::LinearAlgebra::distributed::Vecto
             r<double>; typename std::enable_if<(! dealii::is_serial_vector<T>::value), int>::type <anonymous> = 0; N_Vector = _generic_N_Vector*]':
     1687    /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-src/include/deal.II/sundials/n_vector.templates.h:948:29:   required
             from '_generic_N_Vector* dealii::SUNDIALS::internal::create_empty_nvector() [with VectorType = dealii::LinearAlgebra::distributed::Vector<double>; N_Vector
              = _generic_N_Vector*]'
     1688    /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-src/include/deal.II/sundials/n_vector.templates.h:419:48:   required
             from '_generic_N_Vector* dealii::SUNDIALS::internal::create_nvector(dealii::SUNDIALS::internal::NVectorContent<VectorType>*) [with VectorType = dealii::Lin
             earAlgebra::distributed::Vector<double>; N_Vector = _generic_N_Vector*]'
     1689    /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-src/include/deal.II/sundials/n_vector.templates.h:388:21:   required
             from 'dealii::SUNDIALS::internal::NVectorView<VectorType>::NVectorView(VectorType&) [with VectorType = dealii::LinearAlgebra::distributed::Vector<double>]'
     1690    /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-build-uubbc62/source/sundials/n_vector.inst:74:37:   required from he
             re
  >> 1691    /tmp/root/spack-stage/spack-stage-dealii-9.3.1-uubbc62yai3lwx6meyrl6eodk5av53up/spack-src/include/deal.II/sundials/n_vector.templates.h:534:38: error: inva
             lid conversion from 'MPI_Comm' {aka 'int'} to 'void*' [-fpermissive]
     1692      534 |   return get_communicator<VectorType>(v);
     1693          |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
     1694          |                                      |
     1695          |                                      MPI_Comm {aka int}
```

One possible source of the problem is explained in https://github.com/spack/spack/issues/27183#issuecomment-959787858

> If I had to guess from the error and context, it's that the Deal.II authors assumed that the MPI_Comm is actually a pointer (which is true for OpenMPI) but since you're compiling with MPICH it's actually an int.

**Comment by kronbichler** [comment]
This looks indeed very strange:
https://github.com/dealii/dealii/blob/d98035ee44244d0994598ab7709f0842a97af3ec/include/deal.II/sundials/n_vector.templates.h#L526-L550
Shouldn't we return `MPI_Comm` from those functions and use `MPI_COMM_SELF` for the vectors without MPI, as we do elsewhere. That definition is always around due to https://github.com/dealii/dealii/blob/d98035ee44244d0994598ab7709f0842a97af3ec/include/deal.II/base/mpi.h#L44-L46 @bangerth you worked on this most recently in #12828, can you have a look?

**Comment by bangerth** [comment]
Actually, @sebproell is the right person to ask. I was confused by the function as well when I had reviewed the patch. I don't recall the details, but think that SUNDIALS required this interface?

**Comment by bangerth** [comment]
For reference, the function was added in #11736.

**Comment by bangerth** [comment]
See also https://github.com/dealii/dealii/pull/11701#discussion_r575297184

**Comment by bangerth** [comment]
The underlying issue is that SUNDIALS has an interface that requires using codes to return MPI communicators as a `void*`: See the `nvgetcommunicator` function pointer here: http://runge.math.smu.edu/arkode_dev/doc/guide/build/html/nvectors/NVector_Description.html

But then the SUNDIALS manual at https://computing.llnl.gov/sites/default/files/cv_guide-5.7.0.pdf  on p. 137 documents this function as follows:
```
Returns a pointer to the MPI Comm object associated with the vector (if applicable). For
MPI-unaware vector implementations, this should return NULL.
```
So I think we misunderstood what the function is supposed to do: It's not supposed to return the communicator itself as a `void*`, but it's supposed to return a *pointer to the communicator* or, if the vector is serial, a `nullptr`.

**Comment by sebproell** [comment]
> So I think we misunderstood what the function is supposed to do: It's not supposed to return the communicator itself as a void*, but it's supposed to return a pointer to the communicator or, if the vector is serial, a nullptr.

@bangerth I just checked the implementation of Sundials' own parallel vector and the Trilinos wrapper and you seem to be right about this one :+1:

I will provide a fix.

**Comment by kronbichler** [comment]
Do we also need to fix this in the 9.3.2 release branch, as this was part of the release?

**Comment by bangerth** [comment]
Yes, we need to take this over so that xSDK gets it in the next release (which will use 9.3.2).

@tamiko ?
